File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/service/VisibilityGroupService.java
package com.taurushq.sdk.protect.client.service;

import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import com.taurushq.sdk.protect.client.mapper.ApiExceptionMapper;
import com.taurushq.sdk.protect.client.mapper.UserMapper;
import com.taurushq.sdk.protect.client.mapper.VisibilityGroupMapper;
import com.taurushq.sdk.protect.client.model.ApiException;
import com.taurushq.sdk.protect.client.model.User;
import com.taurushq.sdk.protect.client.model.VisibilityGroup;
import com.taurushq.sdk.protect.openapi.ApiClient;
import com.taurushq.sdk.protect.openapi.api.RestrictedVisibilityGroupsApi;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetUsersByVisibilityGroupIDReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetVisibilityGroupsReply;

import java.util.List;

/**
 * Service for managing visibility groups in the Taurus Protect system.
 * <p>
 * Visibility groups are used to control data access. Users can only see
 * wallets, addresses, and other entities that belong to their assigned
 * visibility groups. This provides fine-grained access control within
 * a tenant.
 * <p>
 * Example usage:
 * <pre>{@code
 * // Get all visibility groups
 * List<VisibilityGroup> groups = client.getVisibilityGroupService().getVisibilityGroups();
 *
 * // Get users in a specific visibility group
 * List<User> users = client.getVisibilityGroupService().getUsersByVisibilityGroup("vg-123");
 * }</pre>
 *
 * @see VisibilityGroup
 */
public class VisibilityGroupService {

    private final RestrictedVisibilityGroupsApi visibilityGroupsApi;
    private final ApiExceptionMapper apiExceptionMapper;

    /**
     * Creates a new VisibilityGroupService.
     *
     * @param apiClient          the API client for making requests
     * @param apiExceptionMapper the mapper for converting API exceptions
     * @throws NullPointerException if any parameter is null
     */
    public VisibilityGroupService(final ApiClient apiClient,
                                   final ApiExceptionMapper apiExceptionMapper) {
        Preconditions.checkNotNull(apiClient, "apiClient must not be null");
        Preconditions.checkNotNull(apiExceptionMapper, "apiExceptionMapper must not be null");
        this.visibilityGroupsApi = new RestrictedVisibilityGroupsApi(apiClient);
        this.apiExceptionMapper = apiExceptionMapper;
    }

    /**
     * Retrieves all visibility groups.
     *
     * @return a list of all visibility groups
     * @throws ApiException if the API call fails
     */
    public List<VisibilityGroup> getVisibilityGroups() throws ApiException {
        try {
            TgvalidatordGetVisibilityGroupsReply reply =
                    visibilityGroupsApi.userServiceGetVisibilityGroups();
            return VisibilityGroupMapper.INSTANCE.fromDTOList(reply.getResult());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves users assigned to a specific visibility group.
     *
     * @param visibilityGroupId the visibility group ID
     * @return a list of users in the visibility group
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if visibilityGroupId is null or empty
     */
    public List<User> getUsersByVisibilityGroup(final String visibilityGroupId) throws ApiException {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(visibilityGroupId),
                "visibilityGroupId must not be null or empty");
        try {
            TgvalidatordGetUsersByVisibilityGroupIDReply reply =
                    visibilityGroupsApi.userServiceGetUsersByVisibilityGroupID(visibilityGroupId);
            return UserMapper.INSTANCE.fromDTO(reply.getResult());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }
}

File: taurus-protect-sdk-typescript/src/services/visibility-group-service.ts
/**
 * Visibility group service for Taurus-PROTECT SDK.
 *
 * Provides methods for managing visibility groups in the Taurus-PROTECT system.
 * Visibility groups are used to control data access. Users can only see
 * wallets, addresses, and other entities that belong to their assigned
 * visibility groups.
 */

import { ValidationError } from '../errors';
import type { RestrictedVisibilityGroupsApi } from '../internal/openapi/apis/RestrictedVisibilityGroupsApi';
import { visibilityGroupsFromDto } from '../mappers/visibility-group';
import { usersFromDto } from '../mappers/user';
import type { VisibilityGroup } from '../models/visibility-group';
import type { User } from '../models/user';
import { BaseService } from './base';

/**
 * Service for managing visibility groups in the Taurus-PROTECT system.
 *
 * Visibility groups are used to control data access. Users can only see
 * wallets, addresses, and other entities that belong to their assigned
 * visibility groups. This provides fine-grained access control within
 * a tenant.
 *
 * @example
 * ```typescript
 * // Get all visibility groups
 * const groups = await client.visibilityGroups.list();
 * for (const group of groups) {
 *   console.log(`${group.name}: ${group.description}`);
 * }
 *
 * // Get users in a specific visibility group
 * const users = await client.visibilityGroups.getUsersByVisibilityGroup('vg-123');
 * for (const user of users) {
 *   console.log(`${user.firstName} ${user.lastName}`);
 * }
 * ```
 */
export class VisibilityGroupService extends BaseService {
  private readonly visibilityGroupsApi: RestrictedVisibilityGroupsApi;

  /**
   * Creates a new VisibilityGroupService instance.
   *
   * @param visibilityGroupsApi - The RestrictedVisibilityGroupsApi instance from the OpenAPI client
   */
  constructor(visibilityGroupsApi: RestrictedVisibilityGroupsApi) {
    super();
    this.visibilityGroupsApi = visibilityGroupsApi;
  }

  /**
   * Lists all visibility groups.
   *
   * @returns Array of visibility groups
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const groups = await visibilityGroupService.list();
   * for (const group of groups) {
   *   console.log(`${group.name}: ${group.userCount} users`);
   * }
   * ```
   */
  async list(): Promise<VisibilityGroup[]> {
    return this.execute(async () => {
      const response = await this.visibilityGroupsApi.userServiceGetVisibilityGroups();

      const result =
        (response as Record<string, unknown>).result ??
        (response as Record<string, unknown>).visibilityGroups;
      return visibilityGroupsFromDto(result as unknown[]);
    });
  }

  /**
   * Gets users assigned to a specific visibility group.
   *
   * @param visibilityGroupId - The visibility group ID
   * @returns Array of users in the visibility group
   * @throws {@link ValidationError} If visibilityGroupId is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const users = await visibilityGroupService.getUsersByVisibilityGroup('vg-123');
   * for (const user of users) {
   *   console.log(`${user.firstName} ${user.lastName} (${user.email})`);
   * }
   * ```
   */
  async getUsersByVisibilityGroup(visibilityGroupId: string): Promise<User[]> {
    if (!visibilityGroupId || visibilityGroupId.trim() === '') {
      throw new ValidationError('visibilityGroupId is required');
    }

    return this.execute(async () => {
      const response =
        await this.visibilityGroupsApi.userServiceGetUsersByVisibilityGroupID({
          visibilityGroupID: visibilityGroupId,
        });

      const result =
        (response as Record<string, unknown>).result ??
        (response as Record<string, unknown>).users;
      return usersFromDto(result as unknown[]);
    });
  }
}
